File: gempak/source/programs/gd/gdgrib/gdigit.c
#include "gdgrib.h"

void gdigit ( const int *ival, const int *ibase, int *ndig, int *idigs,
              int *iret )
/************************************************************************
 * gdigit 								*
 *									*
 * This subroutine computes the individual digits of a decimal input	*
 * number in any arbitrary base.					*
 *									*
 * The digits are ordered beginning with the one's digit; so, IDIGS (1) *
 * is multiplied by IBASE**0 = 1, IDIGS (2) by IBASE**1, and so on, in  *
 * recovering the value in the new base.				*
 *									*
 * gdigit ( ival, ibase, ndig, idigs, iret )				*
 *									*
 * Input parameters:							*
 *	*ival		const int	Input decimal value		*
 *	*ibase		const int	Base for the output digits	*
 *									*
 * Input and output parameter:						*
 *	*ndig		int		Input: max # of digits allowed	*
 *					Output: # of digits needed	*
 *									*
 * Output parameters:							*
 *	*idigs		int		Output digits			*
 *	*iret		int		Return code			*
 *					  0 = normal return		*
 *					 -1 = base cannot be < 2	*
 *					 -2 = not enough digits		*
 *					 -3 = input value < 0		*
 **									*
 * Log:									*
 * K. Brill/HPC		 8/99						*
 * R. Tian/SAIC		 9/06		Recoded from Fortran		*
 ************************************************************************/
{
    int num, idx;
    double dnum;
/*----------------------------------------------------------------------*/
    *iret = 0;

    if ( *ibase < 2 ) {
	*iret = -1;
	return;
    }
    if ( *ival < 0 ) {
	*iret = -3;
	return;
    }
    for ( idx = 0; idx < *ndig; idx++ ) idigs[idx] = 0;

    /*
     * Convert number to new base.
     */
    num = *ival;
    idx = 0;
    while ( num > 0 ) {
        idigs[idx++] = num % (*ibase);
	dnum = floor ( (double)num / (*ibase) );
	num = (int)( dnum );
    }
    *ndig = idx;

    return;
}

File: gempak/source/programs/na/gd2ndfd/gdigit.f
	SUBROUTINE GDIGIT  ( ival, ibase, ndig, idigs, iret )
C************************************************************************
C* GDIGIT 								*
C*									*
C* This subroutine computes the individual digits of a decimal input	*
C* number in any arbitrary base.					*
C*									*
C* The digits are ordered beginning with the one's digit; so, IDIGS (1) *
C* is multiplied by IBASE**0 = 1, IDIGS (2) by IBASE**1, and so on, in  *
C* recovering the value in the new base.				*
C*									*
C* GDIGIT   ( IVAL, IBASE, NDIG, IDIGS, IRET )				*
C*									*
C* Input parameters:							*
C*	IVAL		INTEGER		Input decimal value		*
C*	IBASE		INTEGER		Base for the output digits	*
C*									*
C* Input and output parameter:						*
C*	NDIG		INTEGER		Input: max # of digits allowed	*
C*					Output: # of digits needed	*
C*									*
C* Output parameters:							*
C*	IDIGS (NDIG)	INTEGER		Output digits			*
C*	IRET		INTEGER		Return code			*
C*					  0 = normal return		*
C*					 -1 = base cannot be < 2	*
C*					 -2 = not enough digits		*
C*					 -3 = input value < 0		*
C**									*
C* Log:									*
C* K. Brill/HPC		 8/99						*
C************************************************************************
	INTEGER		idigs (*)
C*
	DOUBLE PRECISION	v, b
C------------------------------------------------------------------------
	IF ( ibase .lt. 2 ) THEN
	    iret = -1
	    RETURN
	END IF
	IF ( ival .lt. 0 ) THEN
	    iret = -3
	    RETURN
	END IF
	iret = 0
	nmax = ndig
	DO i = 1, nmax
	    idigs (i) = 0
	END DO
C*
	v = DFLOAT ( ival )
	b = DFLOAT ( ibase )
	ndig = INT ( DLOG ( v ) / DLOG ( b ) ) + 1
	itest = ibase ** ndig
	IF ( itest .lt. ival ) ndig = ndig + 1
	IF ( ndig .gt. nmax ) THEN
	    iret = -2
	    RETURN
	END IF
	iv = ival
	idig = ndig
	DO WHILE ( idig .gt. 0 )
	    npwr = idig - 1
	    id = ibase ** npwr
	    idigs (idig) = iv / id
	    iv = iv - idigs (idig) * id
	    idig = idig - 1
	END DO
C*
	RETURN
	END
